libp2p/multiaddress.nim

```diff
@@ -816,29 +820,26 @@ proc init*(mtype: typedesc[MultiAddress]): MultiAddress =
   result.data = initVBuffer()
 
 proc init*(mtype: typedesc[MultiAddress],
-           address: IpAddress, protocol: Protocol, port: Port): MaResult[MultiAddress] =
-  ## Initialize MultiAddress using stdlib's net.IpAddress (IPv4/IPv6) and
-  ## net.Protocol (UDP/TCP) information.
-  var res: MultiAddress
-  res.data = initVBuffer()
-  let familyProto = case address.family
-                    of IpAddressFamily.IPv4: getProtocol("ip4")
-                    of IpAddressFamily.IPv6: getProtocol("ip6")
-  let protoProto = case protocol
-                   of IPPROTO_TCP: getProtocol("tcp")
-                   of IPPROTO_UDP: getProtocol("udp")
-                   else: return err("multiaddress: protocol should be either TCP or UDP")
-
-  res.data.write(familyProto.mcodec)
-  if not familyProto.coder.stringToBuffer($address, res.data):
-    err("multiaddress: Error encoding IPv4/IPv6 address")
-  else:
-    res.data.write(protoProto.mcodec)
-    if not protoProto.coder.stringToBuffer($port, res.data):
-      err("multiaddress: Error encoding port number")
-    else:
-      res.data.finish()
-      ok(res)
+           address: ValidIpAddress,
+           protocol: IpTransportProtocol,
+           port: Port): MultiAddress =
+  let
+    familyProto = case address.family
+      of IpAddressFamily.IPv4: getProtocol("ip4")
+      of IpAddressFamily.IPv6: getProtocol("ip6")
+
+    protoProto = case protocol
+      of tcpProtocol: getProtocol("tcp")
+      of udpProtocol: getProtocol("udp")
+
+  var data = initVBuffer()
+  data.write(familyProto.mcodec)
+  doAssert familyProto.coder.stringToBuffer($address, data)
```

Comment:
relying on `doAssert` to call its condition always gives me the shivers...

Reply:
So, why can `stringToBuffer` fail? 

In this case, the mechanical resultification doesn't make the code much easier to reason about.

Reply:
well probably can't in this case, but my guess would be that it was intended as some sort of generic codec mechanism where some some translations can fail, so all of them have to pay the price of error handling - I'm not saying that `doAssert` is wrong in this case, just that putting side effects in the conditional is something that I've trained myself to avoid - even if in nim and `doAssert` it kind of works (until someone comes along with a "-d:reallydisableasserts"

Reply:
Yeah, I got that - it's fixed here:
https://github.com/status-im/nim-libp2p/pull/204/commits/0c57521028eefd99baca79fb82e09f6baf5ed258

Reply:
well, even the mechanical translation makes it easier to reason about because now you see all the places that make or don't make sense, so you know what the next step in refactoring is. 

it's easier to deal with the volume of dumb changes in one commit and do the cleanup in the next, sometimes, or the change snowballs into such absurd amounts of fixes of missing error  handling that it never ends.